[clang-cl][flang][dxc] Fix opts exposed to clang-cl/dxc by mistake

Maetveis opened the pull request:
When these options were enabled for flang the visibility was also extended to clang-cl and dxc. I believe this was due to a misunderstanding of the default value for `Visibility`.

This is marked as draft because I would like to discuss if any of these options should remain, and once that is done add tests to ensure they or are not supported for the relevant driver modes.

Maetveis committed:
[clang-cl][flang][dxc] Fix opts exposed to clang-cl/dxc by mistake

When these options were enabled for flang the visibility was also
extended to clang-cl and dxc. I believe this was due to a misunderstanding
of the default value for `Visibility`.

Maetveis commented:
@tarunprabhu Continuing [our discussion on #108868](https://github.com/llvm/llvm-project/pull/108868/files/612b93eda5695306293bb8ca6e27a810b06adf3d#r1867949328) here. These are what I found, by searching for commits that added both `FlangOption` and `CLOption` to Options.td and were still reachable in the blame of the file. It is possible that this is not all of them.

The default can get quite confusing to track because IIUC the `OptionGroup`'s `Visibility` gets unioned with the visibility defined on the record.

Maetveis commented:
@HaohaiWen can you check if this makes sense for clang-cl? @bob80905 same for DXC please.

tarunprabhu commented:
Thanks for this. I think you have caught all everything I changed. But I'll take another look to see if there was anything else.

tarunprabhu commented:
I checked and I don't think there are any others that have been accidentally enabled. Are there any tests that check to these options produce warnings? If not, it might be worth adding them. What do you think?

HaohaiWen commented on clang/include/clang/Driver/Options.td:
```diff
@@ -1055,11 +1055,11 @@ def z : Separate<["-"], "z">, Flags<[LinkerInput]>,
   HelpText<"Use the new offloading linker to perform the link job.">;
 def Xlinker : Separate<["-"], "Xlinker">, Flags<[LinkerInput, RenderAsInput]>,
-  Visibility<[ClangOption, CLOption, FlangOption, DXCOption]>,
+  Visibility<[ClangOption, FlangOption]>,
```
For clang-cl, /link would pass all args after it to linker and that would cause some trouble when using cmake.
e.g. cmake may use clang-cl as "linker" and append some args after flags specified by CMAKE_EXE_LINKER_FLAGS.
In that case, -Xlinker is better choice.

HaohaiWen commented:
Better to double check with author who specified those CL flags.

Maetveis commented:
> Better to double check with author who specified those CLOption.

Do you mean the person that added `CLOption` to these flags? That was @tarunprabhu, but IIUC that was by mistake; his intention was to extend these flags only to Flang.

Otherwise `-Xlinker` has been around for very long. I guess I will revert these flags to be clang and flang only, then open issues for the discussion of potentially enabling them to clang-cl.

Maetveis committed:
[Clang][Driver][Test] Add test to check unsupported options

For now this is done only for the problematic options but it could
be automatically generated for all driver modes and all unsupported
options by parsing tablegen's json output.

tarunprabhu commented:
> > Better to double check with author who specified those CLOption.
> 
> Do you mean the person that added `CLOption` to these flags? That was @tarunprabhu, but IIUC that was by mistake; his intention was to extend these flags only to Flang.

That is correct, the intention was to extend those flags only to Flang

HaohaiWen commented:
> > Better to double check with author who specified those CLOption.
> 
> Do you mean the person that added `CLOption` to these flags? That was @tarunprabhu, but IIUC that was by mistake; his intention was to extend these flags only to Flang.
> 
> Otherwise `-Xlinker` has been around for very long. I guess I will revert these flags to be clang and flang only, then open issues for the discussion of potentially enabling them to clang-cl.

Some of our project have already used -Xlinker on Windows. Is there any equivalent flag for CL?
Can we keep this flag for CL in this PR and open an issue to discuss whether it should be disabled for CL?

Maetveis committed:
Restore -Xlinker for clang-cl

Maetveis commented:
> Can we keep this flag for CL in this PR and open an issue to discuss whether it should be disabled for CL?

Sure, see latest commit and https://github.com/llvm/llvm-project/issues/119179.

Maetveis commented:
@HaohaiWen can you review and approve if it looks good now for clang-cl? I'll try pinging discord and waiting a little bit for HLSL people before merging.

Gentle ping @bob80905.

HaohaiWen reviewed (approved):
LGTM for -Xlinker.

HaohaiWen commented:
@MaskRay, may need to merge https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/fprofile-sample-use.c to clang/test/Driver/unknown-arg-drivermodes.test

MaskRay commented on clang/test/Driver/unknown-arg-drivermodes.test:
```diff
@@ -0,0 +1,53 @@
+// RUN: -static                    \
+// RUN: -stdlib                    \
+// RUN: -Xoffload-linker           \
+// RUN: -### -x c++ -c - < /dev/null 2>&1 | FileCheck %s --check-prefix=CL
```
need --implicit-check-not=warning:

bob80905 reviewed (approved):
LGTM

Maetveis committed:
[Clang][Driver][Test] Add implicit-check-not

Maetveis commented:
ping @MaskRay

tarunprabhu reviewed (approved):
Thanks for adding the test as well. LGTM.
